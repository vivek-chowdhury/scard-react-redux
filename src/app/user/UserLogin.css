.login-container {
  display: flex;
  flex-direction: column;
  width: 100%;
  height: 100%;
  padding-top: 40px;
  align-items: center;
}
.card {
  position: relative;
  display: flex;
  flex-direction: column;
  min-width: 500px;
  max-width: 600px;
  word-wrap: break-word;
  background-color: #fff;
  background-clip: border-box;
  /* border: 1px solid rgba(0, 0, 0, 0.125); */
  border: 1px solid #007bff;
  border-radius: 0.25rem;
  margin: 10px;
  padding: 10px;
}
