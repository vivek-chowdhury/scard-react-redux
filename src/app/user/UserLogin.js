import React, { useState, useEffect } from "react";
import { connect } from "react-redux";

import LoginForm from "./login-form/LoginForm";
import { toggleRememberMe, validateUser } from "./state/userActions";
import { toggleSpinner } from "./../appAction";

import "./UserLogin.css";

function UserLogin(props) {
  // Contains user state
  const [user, setUser] = useState(props.user);

  //Contains error state
  const [error, setError] = useState("");

  //Contains true if fields are not empty else contain false
  const [isFieldEmpty, setIsFieldEmpty] = useState(true);

  /**
   * @description This method is responsible for checking user detail and if store already has
   * information of previous logged in user then it will pre-populate credentials if
   * user is logged out else navigate user to products screen if user is validated.
   */
  useEffect(() => {
    if (props.user.isLoggedIn) {
      props.history.push("/products");
    }
    if (props.user.error !== "") {
      updateErrorMessage();
    }
  }, [props.history, props.user, props]);

  /**
   * @description This method is responsible for displaying error message
   * if somehow user failed to validate.
   */
  const updateErrorMessage = () => {
    setError("Invalid User name or Password, please try again !");
  };

  /**
   * @description This method is invoked when user start updating login form for validation.
   * It is responsible for updating member variable which application will use for valiation.
   *
   * @param {Element} param Contains reference of target element.
   */
  const handleFieldModified = ({ target }) => {
    setUser({
      ...user,
      [target.name]:
        target.name !== "rememberMe" ? target.value : target.checked,
    });
    if (target.name === "rememberMe") {
      props.toggleRememberMe(target.checked);
    }
    const required = user.username === "" && user.password === "";
    setIsFieldEmpty(required);
  };

  /**
   * @description This method is invoked when user clicks on the submit button,
   * it is responsible for validating user credential by sending user id &
   * password to server for validation.
   * @param {*} event Contains reference of event.
   */
  const handleFormSubmit = (event) => {
    event.preventDefault();
    props.toggleSpinner(true);
    props.validateUser(user).then((response) => {
      if (response && response.message) {
        props.toggleSpinner(false);
      }
    });
  };

  return (
    <div className="login-container">
      <div className="card">
        <LoginForm
          {...user}
          error={error}
          isFieldEmpty={isFieldEmpty}
          onChange={handleFieldModified}
          onSubmit={handleFormSubmit}
        />
      </div>
    </div>
  );
}

/**
 * @description This method retrieve slice of state from store
 * @param {*} state
 */
function mapStateToProps(state) {
  return {
    user: state.user,
  };
}

/**
 * @description This object contains all the actions which Component require to notify reducer for
 * any change.
 */
const mapDispatchToProps = {
  toggleRememberMe,
  validateUser,
  toggleSpinner,
};

export default connect(mapStateToProps, mapDispatchToProps)(UserLogin);
