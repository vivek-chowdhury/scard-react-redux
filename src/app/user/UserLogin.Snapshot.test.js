import React from "react";
import UserLogin from "./UserLogin";
import renderer from "react-test-renderer";
import { user } from "./../../mock/mockDatabase";
import configureMockStore from "redux-mock-store";
import { Provider } from "react-redux";

function renderLoginTree(newState) {
  const mockStore = configureMockStore();
  const state = {
    user: {
      username: "",
      password: "",
      fullName: "",
      isLoggedIn: false,
      rememberMe: false,
      error: "",
    },
    app: { stayHidden: true },
  };
  const finalState = newState ? newState : state;
  const store = mockStore(finalState);

  return renderer.create(
    <Provider store={store}>
      <UserLogin user={user} />
    </Provider>
  );
}

it("it should render empty login form", () => {
  const tree = renderLoginTree();
  expect(tree).toMatchSnapshot();
});
