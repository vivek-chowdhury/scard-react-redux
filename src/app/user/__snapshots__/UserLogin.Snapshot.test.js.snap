// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render empty login form 1`] = `
<div
  className="login-container"
>
  <div
    className="card"
  >
    <form
      className="form-horizontal"
      onSubmit={[Function]}
    >
      
      <div
        className="form-group"
      >
        <label
          className="control-label"
          htmlFor="username"
        >
          User id :
        </label>
        <div
          className=""
        >
          <input
            className="form-control input-field"
            id="username"
            name="username"
            onChange={[Function]}
            placeholder="Enter user id"
            required={true}
            type="text"
            value=""
          />
        </div>
        
      </div>
      <div
        className="form-group"
      >
        <label
          className="control-label"
          htmlFor="password"
        >
          Password :
        </label>
        <div
          className=""
        >
          <input
            className="form-control input-field"
            id="password"
            name="password"
            onChange={[Function]}
            placeholder="Enter password"
            required={true}
            type="password"
            value=""
          />
        </div>
        
      </div>
      <div
        className="action-container"
      >
        <input
          className="btn btn-primary"
          disabled={true}
          type="submit"
          value="Login"
        />
        <label
          htmlFor="rememberMe"
        >
          <input
            checked={false}
            id="rememberMe"
            name="rememberMe"
            onChange={[Function]}
            onClick={[Function]}
            type="checkbox"
          />
          Remember me
        </label>
      </div>
    </form>
  </div>
</div>
`;
